### PR TITLE
fix panic for group by binary type

### PR DESCRIPTION
**Pull request by JCOR11599**
We made a bad type assertion for `sql.StringType`.
Additionally, this fixes a issue where `UnaryExpressions` with `GetFields` would incorrectly throw a functional dependency error with `ONLY_FULL_GROUP_BY` enabled.

Fix for second part of: https://github.com/dolthub/dolt/issues/6179

### CODE REVIEW

File: sql/analyzer/validation_rules.go [hunk 1]
```diff
@@ -320,16 +320,16 @@ func expressionReferencesOnlyGroupBys(groupBys []string, expr sql.Expression) bo
 		case nil, sql.Aggregation, *expression.Literal:
 			return false
 		case *expression.Alias, sql.FunctionExpression:
-			if stringContains(groupBys, expr.String()) {
+			if stringContains(groupBys, strings.ToLower(expr.String())) {
```
Comment (max-hoffman) [inline sql/analyzer/validation_rules.go:323]: might be safe to just always lower the target in `stringContains`, there are other uses of the function that might not be airtight

File: sql/rowexec/agg.go [hunk 2]
```diff
@@ -248,7 +249,10 @@ func groupingKey(
 
 		t, isStringType := expr.Type().(sql.StringType)
 		if isStringType && v != nil {
-			err = t.Collation().WriteWeightString(hash, v.(string))
+			v, err = types.ConvertToString(v, t)
+			if err == nil {
```
Comment (max-hoffman) [inline sql/rowexec/agg.go:253]: should anything happen if err != nil? are we dropping an error or is it impossible for this to map to an aggregation group

**Review by max-hoffman — APPROVED** [review]
lgtm, couple flags for suspicious things. need better aggregation testing and validation but this isn't the place to do it

File: sql/rowexec/agg.go (hunk 2 above)
Comment (JCOR11599) [reply sql/rowexec/agg.go:253]: if `err != nil`, it should fall through to the if [below the else](https://github.com/dolthub/go-mysql-server/pull/1844/files/cc2ddbc8bde618b220c4bebfd45149741dd40e1f#diff-dad61c9d4929b839c2f57e8e0739bc4709607c0f62e4c8737c045592d6a1f046R259), and `return 0, err`

**Pull request merged**